### PR TITLE
Fix for RTTI::RVA, added dumpClassHierarchy 

### PR DESCRIPTION
**Pull request by morcSkyrim**
Commit messages should be pretty self explanatory.

I'm a little confused regarding the RVA stuff, but the fix I've provided works great (at least on my machine).

### CODE REVIEW

**Comment by Ryan-rsm-McKenzie** [comment]
I won't consider this until you run clang-format

**Comment by morcSkyrim** [comment]
I'm also fairly certain that the assembly for the write_Branch6 and write_Branch5 can be simplified as well, and the srcAssembly can be shared between all the write functions.

File: include/SKSE/Trampoline.h [hunk 1]
```diff
@@ -168,10 +168,111 @@ namespace SKSE
...
+			mem->srcOp2 = *reinterpret_cast<std::uint16_t*>(a_src + 4);
+			mem->jmp = 0xE9;
+			mem->disp = static_cast<std::uint32_t>(disp_ret);
+
+			SrcAssembly assembly;
+			assembly.opcode = 0xE9;
+			assembly.disp = static_cast<std::int32_t>(disp);
+			REL::safe_write(a_src, assembly);
+		}
+
```
Comment (Ryan-rsm-McKenzie) [inline include/SKSE/Trampoline.h:275]: Trampoline already contains branching facilities. These new functions are redundant.

File: include/RE/RTTI.cpp [hunk 2]
```diff
@@ -58,16 +58,35 @@ namespace RE
...
 			char buf[0x1000];
 			UnDecorateSymbolName(col->typeDescriptor->name + 1, buf, sizeof(buf), UNDNAME_NO_ARGUMENTS);
 			auto name = buf;
 #else
 			auto name = col->typeDescriptor->name;
 #endif
+			SKSE::log::info("0x{:08X}: {}", a_obj, name);
+		}
 
-			SKSE::log::debug("0x{:08X}: {}", a_obj, name);
```
Comment (Ryan-rsm-McKenzie) [inline include/RE/RTTI.cpp:70]: Please don't make unrelated changes.

File: include/RE/RTTI.cpp [hunk 3]
```diff
@@ -58,16 +58,35 @@ namespace RE
...
+		{
+			uintptr_t* vtbl = *static_cast<std::uintptr_t**>(a_obj) - 1;
+			auto	   col = *reinterpret_cast<RTTI::CompleteObjectLocator**>(vtbl);
+			SKSE::log::debug("0x{:08X}", a_obj);
+			RTTI::BaseClassArray baseClassArray = (col->classDescriptor->baseClassArray);
+			for (uint32_t i = 0; i < col->classDescriptor->numBaseClasses; i++) {
+				RTTI::BaseClassDescriptor*		baseClassDesc = baseClassArray[i];
+				RTTI::RVA<RTTI::TypeDescriptor> testTypeDesc = baseClassDesc->typeDescriptor;
+#ifdef _DEBUG
+				static char buf[0x100];
```
Comment (Ryan-rsm-McKenzie) [inline include/RE/RTTI.cpp:81]: There's no need for this to be static.

File: include/RE/RTTI.h [hunk 4]
```diff
@@ -111,30 +130,8 @@ namespace RE
...
-			// members
-			RVA<TypeDescriptor>						   typeDescriptor;	   // 00
-			std::uint32_t							   numContainedBases;  // 04
-			PMD										   pmd;				   // 08
-			stl::enumeration<Attribute, std::uint32_t> attributes;		   // 14
-		};
-		static_assert(sizeof(BaseClassDescriptor) == 0x18);
-
 		void DumpTypeName(void* a_obj);
+		void DumpClassHier(void* a_obj);
```
Comment (Ryan-rsm-McKenzie) [inline include/RE/RTTI.h:134]: You can inspect a class's hierarchy in ida. I don't see the need for this at runtime.

**Review by Ryan-rsm-McKenzie — CHANGES_REQUESTED** [review]

File: include/SKSE/Trampoline.h (hunk 1 above)
Comment (morcSkyrim) [reply include/SKSE/Trampoline.h:275]: The functions automate the return process, how does anything in the trampoline class provide this functionality?

File: include/RE/RTTI.h (hunk 4 above)
Comment (morcSkyrim) [reply include/RE/RTTI.h:134]: I don't have IDA, and no other open source reverse engineering utilities actually allow for the importing of large number of type definitions making this my only option.

File: include/SKSE/Trampoline.h (hunk 1 above)
Comment (Ryan-rsm-McKenzie) [reply include/SKSE/Trampoline.h:275]: If you need to call the original function, then you can store the address returned by existing facilities and call it later. The method you propose assumes the instrumented call site uses a `call` op, and does not preserve arguments for the injected destination. I will not accept this.

File: include/RE/RTTI.h (hunk 4 above)
Comment (Ryan-rsm-McKenzie) [reply include/RE/RTTI.h:134]: I'd suggest you hand roll a utility for this then.

**Comment by morcSkyrim** [comment]
Should just be the fix to RVA and BaseClassArray now.
